File: Cuate/Providers/DocumentTextQuery.swift
import Foundation

/// The read_document tool's text side: page markers, page ranges, paragraph
/// search and the result cap. Pure Foundation — compiled standalone by
/// `scripts/DocumentContractTest.swift`.
nonisolated enum DocumentTextQuery {
    /// Per tool call. Mirrors the Plaud note cap: big enough for a chapter,
    /// small enough that one call can't crowd out the conversation.
    static let maxResultCharacters = 30_000
    static let maxSearchHits = 20
    static let truncationNote = "[Truncated — request a page range or a query]"

    // MARK: - Page markers

    static func pageMarker(_ number: Int) -> String { "[Page \(number)]" }

    /// Joins per-page texts with markers. Empty pages keep their marker so
    /// the numbering stays aligned with the PDF viewer's.
    static func join(pages: [String]) -> String {
        pages.enumerated().map { index, page in
            pageMarker(index + 1) + "\n" + page.trimmingCharacters(in: .whitespacesAndNewlines)
        }.joined(separator: "\n\n")
    }

    /// Splits marked text back into pages (index 0 = page 1). Text without
    /// markers is a single page; text before the first marker is page 1.
    static func pages(of text: String) -> [String] {
        var pages: [Int: [String]] = [:]
        var current = 1
        var sawMarker = false
        for line in text.components(separatedBy: "\n") {
            if let number = markerNumber(line) {
                current = number
                sawMarker = true
                if pages[current] == nil { pages[current] = [] }
                continue
            }
            pages[current, default: []].append(line)
        }
        guard sawMarker else {
            return [text.trimmingCharacters(in: .whitespacesAndNewlines)]
        }
        let last = pages.keys.max() ?? 1
        return (1...last).map { number in
            (pages[number] ?? []).joined(separator: "\n").trimmingCharacters(in: .whitespacesAndNewlines)
        }
    }

    private static func markerNumber(_ line: String) -> Int? {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        guard trimmed.hasPrefix("[Page "), trimmed.hasSuffix("]") else { return nil }
        let inner = trimmed.dropFirst("[Page ".count).dropLast()
        return Int(inner)
    }

    // MARK: - Page ranges

    /// "3-5", "7", " 2 – 4 " → a closed 1-based range clamped to the
    /// document; nil for garbage or a range that starts past the last page.
    static func parsePageRange(_ raw: String?, pageCount: Int) -> ClosedRange<Int>? {
        guard let raw, pageCount > 0 else { return nil }
        let cleaned = raw
            .replacingOccurrences(of: "–", with: "-")
            .replacingOccurrences(of: "—", with: "-")
            .replacingOccurrences(of: " ", with: "")
        guard !cleaned.isEmpty else { return nil }
        let parts = cleaned.split(separator: "-", omittingEmptySubsequences: false).map(String.init)
        let start: Int
        let end: Int
        switch parts.count {
        case 1:
            guard let single = Int(parts[0]) else { return nil }
            start = single; end = single
        case 2:
            guard let first = Int(parts[0]), let second = Int(parts[1]) else { return nil }
            start = min(first, second); end = max(first, second)
        default:
            return nil
        }
        guard start >= 1, start <= pageCount else { return nil }
        return start...min(end, pageCount)
    }

    // MARK: - Search

    struct Hit: Equatable {
        let page: Int
        let paragraph: String
    }

    /// Case-insensitive paragraph match with page numbers. Plain substring
    /// search: a 40-page contract is a few hundred paragraphs.
    static func search(_ text: String, query: String) -> [Hit] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !needle.isEmpty else { return [] }
        var hits: [Hit] = []
        for (index, page) in pages(of: text).enumerated() {
            for paragraph in page.components(separatedBy: "\n\n") {
                let trimmed = paragraph.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !trimmed.isEmpty, trimmed.lowercased().contains(needle) else { continue }
                hits.append(Hit(page: index + 1, paragraph: trimmed))
                if hits.count >= maxSearchHits { return hits }
            }
        }
        return hits
    }

    // MARK: - Rendering

    /// Applies the cap; the note replaces the tail, never lands mid-word.
    static func capped(_ text: String) -> String {
        guard text.count > maxResultCharacters else { return text }
        var head = String(text.prefix(maxResultCharacters))
        if let cut = head.lastIndex(where: { $0 == "\n" || $0 == " " }) {
            head = String(head[..<cut])
        }
        return head + "\n" + truncationNote
    }

    /// One tool result: a page range, a query, or the whole text — in that
    /// order of precedence when several are given.
    static func render(name: String, text: String, pageRange raw: String?, query: String?) -> String {
        let allPages = pages(of: text)
        var header = "\(name) — \(allPages.count) page\(allPages.count == 1 ? "" : "s")"

        if let raw, !raw.trimmingCharacters(in: .whitespaces).isEmpty {
            guard let range = parsePageRange(raw, pageCount: allPages.count) else {
                return "\(header)\nInvalid page range \"\(raw)\". Valid pages: 1-\(allPages.count)."
            }
            header += ", pages \(range.lowerBound)-\(range.upperBound)"
            let body = range.map { number in
                pageMarker(number) + "\n" + allPages[number - 1]
            }.joined(separator: "\n\n")
            return capped(header + "\n\n" + body)
        }

        if let query, !query.trimmingCharacters(in: .whitespaces).isEmpty {
            let hits = search(text, query: query)
            guard !hits.isEmpty else {
                return "\(header)\nNo matches for \"\(query)\"."
            }
            let lines = hits.map { "Page \($0.page):\n\($0.paragraph)" }.joined(separator: "\n\n")
            let more = hits.count >= maxSearchHits ? "\n[First \(maxSearchHits) matches shown — narrow the query]" : ""
            return capped("\(header), \(hits.count) match\(hits.count == 1 ? "" : "es") for \"\(query)\"\n\n\(lines)\(more)")
        }

        return capped(header + "\n\n" + text)
    }
}

File: android/app/src/main/kotlin/com/aispotlight/android/core/DocumentTextQuery.kt
package com.aispotlight.android.core

/**
 * The read_document tool's text side: page markers, page ranges, paragraph
 * search and the result cap. Port of the Mac `DocumentTextQuery`; pure
 * Kotlin, covered by the unit test with the same cases.
 */
object DocumentTextQuery {
    const val MAX_RESULT_CHARACTERS = 30_000
    const val MAX_SEARCH_HITS = 20
    const val TRUNCATION_NOTE = "[Truncated — request a page range or a query]"

    fun pageMarker(number: Int) = "[Page $number]"

    /** Joins per-page texts with markers; empty pages keep their marker. */
    fun join(pages: List<String>): String =
        pages.mapIndexed { index, page -> pageMarker(index + 1) + "\n" + page.trim() }
            .joinToString("\n\n")

    /** Splits marked text back into pages (index 0 = page 1). */
    fun pages(text: String): List<String> {
        val pages = mutableMapOf<Int, MutableList<String>>()
        var current = 1
        var sawMarker = false
        for (line in text.split("\n")) {
            val number = markerNumber(line)
            if (number != null) {
                current = number
                sawMarker = true
                pages.getOrPut(current) { mutableListOf() }
                continue
            }
            pages.getOrPut(current) { mutableListOf() }.add(line)
        }
        if (!sawMarker) return listOf(text.trim())
        val last = pages.keys.maxOrNull() ?: 1
        return (1..last).map { number -> (pages[number] ?: emptyList()).joinToString("\n").trim() }
    }

    private fun markerNumber(line: String): Int? {
        val trimmed = line.trim()
        if (!trimmed.startsWith("[Page ") || !trimmed.endsWith("]")) return null
        return trimmed.removePrefix("[Page ").removeSuffix("]").toIntOrNull()
    }

    /** "3-5", "7", " 2 – 4 " → a closed 1-based range clamped to the document. */
    fun parsePageRange(raw: String?, pageCount: Int): IntRange? {
        if (raw == null || pageCount <= 0) return null
        val cleaned = raw.replace("–", "-").replace("—", "-").replace(" ", "")
        if (cleaned.isEmpty()) return null
        val parts = cleaned.split("-")
        val start: Int
        val end: Int
        when (parts.size) {
            1 -> { val single = parts[0].toIntOrNull() ?: return null; start = single; end = single }
            2 -> {
                val first = parts[0].toIntOrNull() ?: return null
                val second = parts[1].toIntOrNull() ?: return null
                start = minOf(first, second); end = maxOf(first, second)
            }
            else -> return null
        }
        if (start < 1 || start > pageCount) return null
        return start..minOf(end, pageCount)
    }

    data class Hit(val page: Int, val paragraph: String)

    /** Case-insensitive paragraph match with page numbers (plain substring). */
    fun search(text: String, query: String): List<Hit> {
        val needle = query.trim().lowercase()
        if (needle.isEmpty()) return emptyList()
        val hits = mutableListOf<Hit>()
        pages(text).forEachIndexed { index, page ->
            for (paragraph in page.split("\n\n")) {
                val trimmed = paragraph.trim()
                if (trimmed.isEmpty() || !trimmed.lowercase().contains(needle)) continue
                hits.add(Hit(index + 1, trimmed))
                if (hits.size >= MAX_SEARCH_HITS) return hits
            }
        }
        return hits
    }

    /** Applies the cap; the note replaces the tail, never mid-word. */
    fun capped(text: String): String {
        if (text.length <= MAX_RESULT_CHARACTERS) return text
        var head = text.take(MAX_RESULT_CHARACTERS)
        val cut = head.lastIndexOfAny(charArrayOf('\n', ' '))
        if (cut > 0) head = head.substring(0, cut)
        return head + "\n" + TRUNCATION_NOTE
    }

    /** One tool result: a page range, a query, or the whole text. */
    fun render(name: String, text: String, pageRange: String?, query: String?): String {
        val allPages = pages(text)
        var header = "$name — ${allPages.size} page${if (allPages.size == 1) "" else "s"}"
        if (!pageRange.isNullOrBlank()) {
            val range = parsePageRange(pageRange, allPages.size)
                ?: return "$header\nInvalid page range \"$pageRange\". Valid pages: 1-${allPages.size}."
            header += ", pages ${range.first}-${range.last}"
            val body = range.joinToString("\n\n") { number -> pageMarker(number) + "\n" + allPages[number - 1] }
            return capped("$header\n\n$body")
        }
        if (!query.isNullOrBlank()) {
            val hits = search(text, query)
            if (hits.isEmpty()) return "$header\nNo matches for \"$query\"."
            val lines = hits.joinToString("\n\n") { "Page ${it.page}:\n${it.paragraph}" }
            val more = if (hits.size >= MAX_SEARCH_HITS) "\n[First $MAX_SEARCH_HITS matches shown — narrow the query]" else ""
            return capped("$header, ${hits.size} match${if (hits.size == 1) "" else "es"} for \"$query\"\n\n$lines$more")
        }
        return capped("$header\n\n$text")
    }
}
